core/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -569,6 +568,21 @@ public void shardInitialized(ShardRouting unassignedShard, ShardRouting initiali
             }
         }
 
+        @Override
+        public void unassignedInfoUpdated(ShardRouting unassignedShard, UnassignedInfo newUnassignedInfo) {
+            if (unassignedShard.primary()) {
+                RecoverySource recoverySource = unassignedShard.recoverySource();
+                if (recoverySource.getType() == RecoverySource.Type.SNAPSHOT) {
+                    if (newUnassignedInfo.getLastAllocationStatus() == UnassignedInfo.AllocationStatus.DECIDERS_NO) {
+                        Snapshot snapshot = ((SnapshotRecoverySource) recoverySource).snapshot();
+                        String reason = "shard was denied allocation by all allocation deciders";
+                        changes(snapshot).unassignedShards.put(unassignedShard.shardId(),
```

Comment:
I think instead of adding more types here (unassignedShards), better we do the reverse and fold failedShards, startedShards and unassignedShards into just "updates".
It's not worth separating them just to have this one assertion I've put there.

Reply:
Pff I should have seen that... I agree, that would be better, thanks :)